nano/node/confirmation_height_processor.cpp

```diff
@@ -0,0 +1,238 @@
+		}
+
+		auto transaction (store.tx_begin_read ());
+		auto block_height = (store.block_account_height (transaction, current));
+		nano::account account (ledger.account (transaction, current));
+		release_assert (!store.account_get (transaction, account, account_info));
+		auto confirmation_height = account_info.confirmation_height;
+		auto count_before_open_receive = open_receive_source_pairs.size ();
+
+		auto hash (current);
+		if (block_height > confirmation_height)
+		{
+			collect_unconfirmed_receive_and_sources_for_account (block_height, confirmation_height, current, genesis_hash, open_receive_source_pairs, account, transaction);
+		}
+
+		// If this adds no more open_receive blocks, then we can now confirm this account as well as the linked open/receive block
+		// Collect as pending any writes to the database and do them in bulk after a certain time.
+		auto confirmed_receives_pending = (count_before_open_receive != open_receive_source_pairs.size ());
+		if (!confirmed_receives_pending)
+		{
+			if (block_height > confirmation_height)
+			{
+				update_confirmation_height (account, { hash, block_height }, pending);
+			}
+
+			if (open_receive_details)
+			{
+				update_confirmation_height (open_receive_details->account, open_receive_details->block_hash_height_pair, pending);
+			}
+
+			if (!open_receive_source_pairs.empty ())
+			{
+				open_receive_source_pairs.pop ();
+			}
+		}
+
+		// Check whether writing to the database should be done now
+		if ((timer.after_deadline (std::chrono::milliseconds (50)) || open_receive_source_pairs.empty ()) && !pending.empty ())
+		{
+			write_pending (pending);
+			timer.restart ();
+		}
+	} while (!open_receive_source_pairs.empty ());
+}
+
+void nano::confirmation_height_processor::update_confirmation_height (nano::account const & account_a, block_hash_height_pair const & block_hash_height_pair_a, std::unordered_map<nano::account, block_hash_height_pair> & pending)
+{
+	auto it = pending.find (account_a);
+	if (it != pending.end ())
+	{
+		if (block_hash_height_pair_a.height > it->second.height)
+		{
+			it->second = block_hash_height_pair_a;
+		}
+	}
+	else
+	{
+		pending.emplace (account_a, block_hash_height_pair_a);
+	}
+}
+
+void nano::confirmation_height_processor::write_pending (std::unordered_map<nano::account, block_hash_height_pair> & all_pending)
+{
+	nano::account_info account_info;
+	auto transaction (store.tx_begin_write ());
+	for (auto & pending : all_pending)
+	{
+		auto error = store.account_get (transaction, pending.first, account_info);
+		release_assert (!error);
+		if (pending.second.height > account_info.confirmation_height)
+		{
+			// Check that the confirmation height and block hash still match, as there
+			// may have been changes outside this processor
+			nano::block_sideband sideband;
+			auto block = store.block_get (transaction, pending.second.hash, &sideband);
+			release_assert (block != nullptr);
+			release_assert (sideband.height == pending.second.height);
```

Comment:
This release_assert hit after a few minutes on a live ledger that was almost in sync.

`sideband.height` == 10
`pending.second.height` == 11

Block hash 66B06DDC0D40AC86522CEEF5747FDFE7DD503B1D108B0B00432FB5E714F8BEC5

Reply:
The off by one appears to originate from the `confirmation_height + num_to_confirm` logic, which was done as an optimization to prevent having to get the sideband data from the block. I can't think why it would be an issue in some cases though, but I'm just going to use the sideband data directly. I'm also doing another pessimistic approach as in-between the reads and writes of the confirmation heights the chains could be altered possibly forking and removing blocks I was in the process of confirming, or perhaps this is a non-issue?